File: samples/slam/visual_odometry.cpp
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
// Copyright (C) 2026, BigVision LLC, all rights reserved.

#include <opencv2/ptcloud.hpp>
#include <opencv2/features.hpp>
#include <opencv2/core.hpp>
#include <opencv2/core/quaternion.hpp>
#include <opencv2/core/utils/filesystem.hpp>
#include <opencv2/core/utils/logger.hpp>
#include <opencv2/imgcodecs.hpp>

#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>

#include "../dnn/common.hpp"

using namespace cv;
using namespace std;

const string about =
    "Modular monocular visual odometry: feature tracking + PnP, no bundle adjustment\n"
    "or loop closure.\n\n"
    "Feeds a directory of images through a pluggable feature detector + matcher pair,\n"
    "estimates the camera trajectory, and writes the result to --output in COLMAP text\n"
    "format. This sample wires up one ONNX detector/matcher pair by default; the\n"
    "pipeline (cv::slam::VisualOdometry) accepts any cv::Feature2D + cv::DescriptorMatcher.\n"
    "To run:\n"
    "\t ./example_slam_visual_odometry --aliked=aliked.onnx --lightglue=lightglue.onnx --images=./seq\n"
    "Sample command (run on the GPU):\n"
    "\t ./example_slam_visual_odometry --aliked=aliked.onnx --lightglue=lightglue.onnx --images=./seq --target=cuda\n";

const string param_keys =
    "{ help h           |        | Print help message }"
    "{ aliked           | <none> | Path to detector ONNX model }"
    "{ lightglue        | <none> | Path to matcher ONNX model }"
    "{ images           | <none> | Path to directory with input images }"
    "{ output           | vo_out | Output directory for trajectory and map }"
    "{ fx               | 718.856  | Camera focal length X }"
    "{ fy               | 718.856  | Camera focal length Y }"
    "{ cx               | 607.1928 | Camera principal point X }"
    "{ cy               | 185.2157 | Camera principal point Y }"
    "{ dist             |        | Lens distortion coeffs, comma-separated k1,k2,p1,p2[,k3,...] (default: none) }"
    "{ progress         | true   | Print per-frame progress logs to the console as they happen }";

const string backend_keys = format(
    "{ backend          | default | Choose one of computation backends: "
                              "default: automatically (by default), "
                              "openvino: Intel's Deep Learning Inference Engine (https://software.intel.com/openvino-toolkit), "
                              "opencv: OpenCV implementation, "
                              "vkcom: VKCOM, "
                              "cuda: CUDA, "
                              "webnn: WebNN }");

const string target_keys = format(
    "{ target           | cpu | Choose one of target computation devices: "
                              "cpu: CPU target (by default), "
                              "opencl: OpenCL, "
                              "opencl_fp16: OpenCL fp16 (half-float precision), "
                              "vpu: VPU, "
                              "vulkan: Vulkan, "
                              "cuda: CUDA, "
                              "cuda_fp16: CUDA fp16 (half-float preprocess) }");

string keys = param_keys + backend_keys + target_keys;

static vector<String> collectImages(const String& imagesDir)
{
    vector<String> allFiles;
    try
    {
        glob(imagesDir, allFiles, false);
    }
    catch (const Exception& e)
    {
        cerr << "glob failed on " << imagesDir << ": " << e.what() << endl;
        return {};
    }

    vector<String> imageFiles;
    imageFiles.reserve(allFiles.size());
    for (const String& file : allFiles)
        if (haveImageReader(file))
            imageFiles.push_back(file);

    sort(imageFiles.begin(), imageFiles.end());
    return imageFiles;
}

static Mat parseDistCoeffs(const String& text)
{
    stringstream stream(text);
    vector<double> coeffs;
    String token;
    while (getline(stream, token, ','))
    {
        const size_t begin = token.find_first_not_of(" \t");
        const size_t end   = token.find_last_not_of(" \t");
        if (begin == String::npos) continue;
        coeffs.push_back(stod(token.substr(begin, end - begin + 1)));
    }
    return coeffs.empty() ? Mat() : Mat(coeffs, true).reshape(1, 1);
}

static bool writeColmapFiles(const Ptr<slam::VisualOdometry>& vo,
                             const Matx33d& K, const Mat& distCoeffs, Size imageSize,
                             const vector<String>& poseImageNames,
                             const String& outputFolder)
{
    if (!utils::fs::createDirectories(outputFolder))
    {
        cerr << "cannot create output directory " << outputFolder << endl;
        return false;
    }

    // camera.txt
    {
        ofstream file(utils::fs::join(outputFolder, "camera.txt").c_str());
        if (!file.is_open()) { cerr << "cannot write camera.txt" << endl; return false; }
        file << "# Camera list with one line of data per camera:\n"
             << "#   CAMERA_ID, MODEL, WIDTH, HEIGHT, PARAMS[fx, fy, cx, cy, dist...]\n";
        file << setprecision(9);
        file << "1 " << (distCoeffs.empty() ? "PINHOLE" : "FULL_OPENCV") << " "
             << imageSize.width << " " << imageSize.height << " "
             << K(0, 0) << " " << K(1, 1) << " " << K(0, 2) << " " << K(1, 2);
        for (int i = 0; i < (int)distCoeffs.total(); ++i)
            file << " " << distCoeffs.at<double>(i);
        file << "\n";
    }

    // images.txt
    {
        const vector<Matx44d>& trajectory = vo->getTrajectory();
        ofstream file(utils::fs::join(outputFolder, "images.txt").c_str());
        if (!file.is_open()) { cerr << "cannot write images.txt" << endl; return false; }
        file << "# Image list with one line of data per image:\n"
             << "#   IMAGE_ID, QW, QX, QY, QZ, TX, TY, TZ, CAMERA_ID, NAME\n"
             << "# Number of images: " << trajectory.size() << "\n";
        file << setprecision(9);
        for (size_t i = 0; i < trajectory.size(); ++i)
        {
            const Matx44d& poseCw = trajectory[i];
            const Quatd q = Quatd::createFromRotMat(poseCw.get_minor<3, 3>(0, 0));
            // COLMAP expects a name relative to the image directory, not a full path.
            const String name = i < poseImageNames.size()
                              ? poseImageNames[i].substr(poseImageNames[i].find_last_of("/\\") + 1)
                              : format("pose_%zu", i);
            file << i << " " << q.w << " " << q.x << " " << q.y << " " << q.z << " "
                 << poseCw(0, 3) << " " << poseCw(1, 3) << " " << poseCw(2, 3)
                 << " 1 " << name << "\n";
        }
    }

    // point3d.txt
    {
        ofstream file(utils::fs::join(outputFolder, "point3d.txt").c_str());
        if (!file.is_open()) { cerr << "cannot write point3d.txt" << endl; return false; }
        file << "# 3D point list with one line of data per point:\n"
             << "#   POINT3D_ID, X, Y, Z, TRACK_LENGTH\n";
        file << setprecision(9);
        for (const slam::MapPoint* point : vo->getMap().mapPoints())
        {
            if (!point || point->bad) continue;
            file << point->id << " "
                 << point->pos.x << " " << point->pos.y << " " << point->pos.z << " "
                 << point->observations.size() << "\n";
        }
    }

    return true;
}

int main(int argc, char** argv)
{
    CommandLineParser parser(argc, argv, keys);
    parser.about(about);

    if (parser.has("help"))
    {
        parser.printMessage();
        return 0;
    }

    const String alikedPath = parser.get<String>("aliked");
    const String lightgluePath = parser.get<String>("lightglue");
    const String imagesDir = parser.get<String>("images");
    const int backendId = getBackendID(parser.get<String>("backend"));
    const int targetId = getTargetID(parser.get<String>("target"));

    if (!parser.check() || alikedPath == "<none>" || lightgluePath == "<none>" || imagesDir == "<none>")
    {
        parser.printErrors();
        parser.printMessage();
        return 1;
    }

    const bool showProgress = parser.get<bool>("progress");
    if (showProgress)
        utils::logging::setLogLevel(utils::logging::LOG_LEVEL_INFO);

    const String outputDir = parser.get<String>("output");

    const Matx33d K(parser.get<double>("fx"), 0., parser.get<double>("cx"),
                    0., parser.get<double>("fy"), parser.get<double>("cy"),
                    0., 0., 1.);
    const Mat distCoeffs = parseDistCoeffs(parser.get<String>("dist"));

    const vector<String> imageFiles = collectImages(imagesDir);
    if (imageFiles.empty())
    {
        cerr << "no images found in " << imagesDir << endl;
        return 1;
    }

    ALIKED::Params detParams;
    detParams.inputSize = Size(640, 640);
    detParams.engine    = dnn::ENGINE_OPENCV;
    detParams.backend   = backendId;
    detParams.target    = targetId;
    auto detector = ALIKED::create(alikedPath, detParams);

    auto matcher = LightGlueMatcher::create(lightgluePath, 0.0f, backendId, targetId);

    // tracking parameters
    slam::OdometryParams voParams;
    voParams.minInitParallaxDeg = 1.5;
    voParams.minInitPoints      = 50;
    voParams.pnpReprojThresh    = 4.0;
    voParams.kfMaxFrames        = 30;
    voParams.localMapTopK       = 10;
    voParams.poseOptEnable      = false;
    voParams.localBaEnable      = false;
    voParams.globalBaEnable     = false;
    voParams.loopEnable         = false;
    voParams.loopCloseEnable    = false;

    auto vo = slam::VisualOdometry::create(detector, matcher, Mat(K), distCoeffs, voParams);

    cout << "images folder : " << imagesDir << "\n"
         << "output folder : " << outputDir << "\n"
         << "images found  : " << imageFiles.size() << "\n\n"
         << "Running feature tracking (PnP), no BA or loop closure.\n"
         << (showProgress
                ? "Per-frame progress is printed below (OpenCV logs to stderr). Pass --progress=false to silence it.\n\n"
                : "Per-frame progress is disabled. Re-run with --progress=true (the default) to see it.\n\n");

    vector<String> poseImageNames;
    String refImageName;
    Size imageSize;
    size_t previousPoseCount = 0;

    const int64 ticksBefore = getTickCount();
    for (size_t i = 0; i < imageFiles.size(); ++i)
    {
        const Mat image = imread(imageFiles[i]);
        if (image.empty())
        {
            cerr << "[frame " << i << "] imread failed: " << imageFiles[i] << endl;
            continue;
        }
        imageSize = image.size();

        const slam::OdometryState stateBefore = vo->getState();
        vo->processFrame(image);
        const slam::OdometryState stateAfter = vo->getState();

        if (stateBefore == slam::NOT_INITIALIZED ||
            (stateBefore == slam::TRACKING && stateAfter == slam::INITIALIZING))
            refImageName = imageFiles[i];

        const size_t poseCount = vo->getTrajectory().size();
        if (poseCount == previousPoseCount + 1)
            poseImageNames.push_back(imageFiles[i]);
        else if (poseCount == previousPoseCount + 2)
        {
            poseImageNames.push_back(refImageName);
            poseImageNames.push_back(imageFiles[i]);
        }
        previousPoseCount = poseCount;
    }
    const double elapsed = (getTickCount() - ticksBefore) / getTickFrequency();

    const bool ok = !vo->getTrajectory().empty();
    bool exported = false;
    if (ok && !outputDir.empty())
    {
        try
        {
            exported = writeColmapFiles(vo, K, distCoeffs, imageSize, poseImageNames, outputDir);
        }
        catch (const Exception& e)
        {
            cerr << "writing output failed: " << e.what() << endl;
        }
    }

    cout << "\n"
         << "================ Visual Odometry Result ================\n"
         << "status        : " << (ok ? "OK" : "FAILED") << "\n"
         << "camera poses  : " << vo->getTrajectory().size() << "\n"
         << "keyframes     : " << vo->getNumKeyframes() << "\n"
         << "map points    : " << vo->getNumMapPoints() << "\n"
         << "elapsed time  : " << fixed << setprecision(2) << elapsed << " s\n";
    if (exported)
        cout << "output        : " << outputDir << "/{camera,images,point3d}.txt\n";
    cout << "==========================================================\n";

    return ok ? 0 : 1;
}

File: samples/slam/visual_odometry.py
'''
Monocular visual odometry with cv.slam.VisualOdometry (ALIKED + LightGlue).

Example:
    python visual_odometry.py --aliked aliked.onnx --lightglue lg.onnx --images ./seq
'''

import argparse
import glob
import os
import sys
import time
import numpy as np
import cv2 as cv


def build_K(fx, fy, cx, cy):
    return np.array([[fx, 0., cx],
                     [0., fy, cy],
                     [0., 0., 1.]], dtype=np.float64)


def list_image_files(images_dir):
    files = [f for f in sorted(glob.glob(os.path.join(images_dir, '*')))
             if cv.haveImageReader(f)]
    return files


def rotation_matrix_to_quaternion(R):
    # Shepperd's method: numerically stable for all rotations.
    trace = R[0, 0] + R[1, 1] + R[2, 2]
    if trace > 0:
        s = np.sqrt(trace + 1.0) * 2
        qw = 0.25 * s
        qx = (R[2, 1] - R[1, 2]) / s
        qy = (R[0, 2] - R[2, 0]) / s
        qz = (R[1, 0] - R[0, 1]) / s
    elif R[0, 0] > R[1, 1] and R[0, 0] > R[2, 2]:
        s = np.sqrt(1.0 + R[0, 0] - R[1, 1] - R[2, 2]) * 2
        qw = (R[2, 1] - R[1, 2]) / s
        qx = 0.25 * s
        qy = (R[0, 1] + R[1, 0]) / s
        qz = (R[0, 2] + R[2, 0]) / s
    elif R[1, 1] > R[2, 2]:
        s = np.sqrt(1.0 + R[1, 1] - R[0, 0] - R[2, 2]) * 2
        qw = (R[0, 2] - R[2, 0]) / s
        qx = (R[0, 1] + R[1, 0]) / s
        qy = 0.25 * s
        qz = (R[1, 2] + R[2, 1]) / s
    else:
        s = np.sqrt(1.0 + R[2, 2] - R[0, 0] - R[1, 1]) * 2
        qw = (R[1, 0] - R[0, 1]) / s
        qx = (R[0, 2] + R[2, 0]) / s
        qy = (R[1, 2] + R[2, 1]) / s
        qz = 0.25 * s
    return qw, qx, qy, qz


def write_colmap_files(vo, K, image_size, pose_filenames, output_dir):
    # Note: cv.slam.Map isn't wrapped for Python (only aggregate counts are,
    # via getNumKeyframes/getNumMapPoints), so only camera intrinsics and the
    # trajectory (images.txt) can be exported here; point3d.txt (map points)
    # requires the C++ sample.
    os.makedirs(output_dir, exist_ok=True)

    with open(os.path.join(output_dir, 'camera.txt'), 'w') as f:
        f.write(f"fx {K[0, 0]:.4f}\n")
        f.write(f"fy {K[1, 1]:.4f}\n")
        f.write(f"cx {K[0, 2]:.4f}\n")
        f.write(f"cy {K[1, 2]:.4f}\n")
        f.write(f"width {image_size[0]}\n")
        f.write(f"height {image_size[1]}\n")

    traj = vo.getTrajectory()
    with open(os.path.join(output_dir, 'images.txt'), 'w') as f:
        f.write("# Image list with two lines of data per image:\n")
        f.write("#   IMAGE_ID, QW, QX, QY, QZ, TX, TY, TZ, CAMERA_ID, NAME\n")
        f.write("#   POINTS2D[] as (X, Y, POINT3D_ID)\n")
        f.write(f"# Number of images: {len(traj)}, mean observations per image: 0.0\n")
        for i, T in enumerate(traj):
            qw, qx, qy, qz = rotation_matrix_to_quaternion(T[:3, :3])
            name = (os.path.basename(pose_filenames[i]) if i < len(pose_filenames)
                    else f"pose_{i}")
            f.write(f"{i} {qw:.6f} {qx:.6f} {qy:.6f} {qz:.6f} "
                    f"{T[0,3]:.6f} {T[1,3]:.6f} {T[2,3]:.6f} 1 {name}\n")


def main():
    parser = argparse.ArgumentParser(
        description='Monocular visual odometry using ALIKED + LightGlue')
    parser.add_argument('--aliked',    required=True,
                        help='Path to ALIKED ONNX model')
    parser.add_argument('--lightglue', required=True,
                        help='Path to LightGlue ONNX model')
    parser.add_argument('--images',    required=True,
                        help='Path to directory with input images')
    parser.add_argument('--output',    default='vo_out',
                        help='Output directory for trajectory and map (default: vo_out)')
    parser.add_argument('--fx', type=float, default=718.856,
                        help='Camera focal length X (default: KITTI-00)')
    parser.add_argument('--fy', type=float, default=718.856,
                        help='Camera focal length Y (default: KITTI-00)')
    parser.add_argument('--cx', type=float, default=607.1928,
                        help='Camera principal point X (default: KITTI-00)')
    parser.add_argument('--cy', type=float, default=185.2157,
                        help='Camera principal point Y (default: KITTI-00)')
    parser.add_argument('--min-parallax', type=float, default=1.5,
                        help='Minimum initialisation parallax in degrees (default: 1.5)')
    parser.add_argument('--min-points', type=int, default=50,
                        help='Minimum initialisation map points (default: 50)')
    args = parser.parse_args()

    det_params = cv.ALIKED.Params()
    det_params.inputSize = (640, 640)
    det_params.engine    = cv.dnn.ENGINE_OPENCV
    detector = cv.ALIKED.create(args.aliked, det_params)

    matcher = cv.LightGlueMatcher.create(
        args.lightglue, 0.0,
        cv.dnn.DNN_BACKEND_DEFAULT,
        cv.dnn.DNN_TARGET_CPU)

    vo_params = cv.slam.OdometryParams()
    vo_params.minInitParallaxDeg = args.min_parallax
    vo_params.minInitPoints      = args.min_points

    K = build_K(args.fx, args.fy, args.cx, args.cy)

    vo = cv.slam.VisualOdometry.create(
        detector, matcher, K, np.array([]), vo_params)

    image_files = list_image_files(args.images)
    if not image_files:
        print(f"no images found in {args.images}", file=sys.stderr)
        return 1

    print(f"images_folder = {args.images}")
    print(f"output_folder = {args.output}")
    print(f"found {len(image_files)} image(s)")

    # Tracks which input image each emitted trajectory pose came from, for images.txt.
    pose_filenames = []
    prev_traj_len = 0
    ref_filename = None
    image_size = (0, 0)
    n_emitted = 0

    t0 = time.perf_counter()
    for i, path in enumerate(image_files):
        img = cv.imread(path)
        if img is None:
            print(f"[FRAME {i}] file={path} imread failed", file=sys.stderr)
            continue
        image_size = (img.shape[1], img.shape[0])

        before = vo.getState()
        emitted = vo.processFrame(img)
        after = vo.getState()
        if emitted:
            n_emitted += 1

        # Track which input image maps to each trajectory pose.
        if before == cv.slam.NOT_INITIALIZED or \
           (before == cv.slam.TRACKING and after == cv.slam.INITIALIZING):
            ref_filename = path

        traj_len = len(vo.getTrajectory())
        added = traj_len - prev_traj_len
        if added == 1:
            pose_filenames.append(path)
        elif added == 2:
            pose_filenames.append(ref_filename)
            pose_filenames.append(path)
        prev_traj_len = traj_len

        print(f"[FRAME {i}] file={path}"
              f" emitted={'yes' if emitted else 'no'}"
              f" keyframes={vo.getNumKeyframes()}"
              f" map_points={vo.getNumMapPoints()}")
    elapsed = time.perf_counter() - t0
    ok = n_emitted > 0

    if ok and args.output:
        write_colmap_files(vo, K, image_size, pose_filenames, args.output)

    print(f"run={'ok' if ok else 'FAILED'}  frames={len(vo.getTrajectory())}  elapsed={elapsed:.2f}s")
    print(f"output -> {args.output}")
    return 0 if ok else 1


if __name__ == '__main__':
    sys.exit(main())
